lib/channels/src/index.ts

```diff
@@ -0,0 +1,152 @@
+export type ChannelHandler<TEventArgs = any> = (event: ChannelEvent<TEventArgs>) => void;
+
+export interface ChannelTransport<TEventArgs = any> {
+  send(event: ChannelEvent<TEventArgs>): void;
+  setHandler(handler: ChannelHandler<TEventArgs>): void;
+}
+
+export interface ChannelEvent<TEventArgs = any> {
+  type: string; // eventName
+  from: string;
+  args: TEventArgs;
+}
+
+export interface Listener<TEventArgs = any> {
+  (...args: TEventArgs[]): void;
+
+  ignorePeer?: boolean;
+}
+
+interface EventsKeyValue {
+  [key: string]: Listener[];
+}
+
+interface ChannelArgs {
+  transport?: ChannelTransport;
+  async?: boolean;
+}
+
+const generateRandomId = () => {
+  // generates a random 13 character string
+  return Math.random()
+    .toString(16)
+    .slice(2);
+};
+
+export class Channel {
+  readonly isAsync: boolean;
+
+  private _sender = generateRandomId();
+  private _events: EventsKeyValue = {};
+  private readonly _transport: ChannelTransport;
+
+  constructor({ transport, async = false }: ChannelArgs = {}) {
+    this.isAsync = async;
+    if (transport) {
+      this._transport = transport;
+      this._transport.setHandler(event => this._handleEvent(event));
+    }
+  }
+
+  get hasTransport() {
+    return !!this._transport;
```

Comment:
I personally prefer the following way of enforcing the return of a boolean based on the existence of an object due to readability. this is minor though. Choose whichever you prefer :)

```suggestion
    return Boolean(this._transport);
```

Reply:
I prefer `!!` too